FEA Add multiclass support to `average_precision_score`

awinml opened the pull request:
#### Reference Issues/PRs
Fixes #17388 
Fixes #17104 

#### What does this implement/fix? Explain your changes.
- Added multi class support to `average_precision_score` by reusing the multi label case and computing it in a OVR manner.
- Update user guide for model evaluation to inform user about multi class support for `average_precision_score`.
- Added multi class example in API documentation. 
- Modified `test_precision_recall_display_validation` in `test_precision_recall_display.py` to fail when predictions for all classes are not supplied to `average_precision_score`.
- Modified `test_average_precision_score_non_binary_class` to fail while computing `average_precision_score` for multi class - multi output case.
- Added `test_average_precision_score_multiclass_tied_values` to test for tied values of predictions when computing `average_precision_score` for the multi class case.
- Added `test_average_precision_score_multiclass_duplicate_values` for duplicate values of y_true and y_score when computing `average_precision_score` for the multi class case.
- Added `test_average_precision_score_multiclass_pos_label_errors` to check correct pos_label argument.

gbolmier committed:
Add multiclass support to average_precision_score

# More detailed explanatory text, if necessary. Wrap it to about 72
# characters or so. In some contexts, the first line is treated as the
# subject of the commit and the rest of the text as the body. The
# blank line separating the summary from the body is critical (unless
# you omit the body entirely); various tools like `log`, `shortlog`
# and `rebase` can get confused if you run the two together.

# Explain the problem that this commit is solving. Focus on why you
# are making this change as opposed to how (the code explains that).
# Are there side effects or other unintuitive consequences of this
# change? Here's the place to explain them.

# Further paragraphs come after blank lines.

#  - Bullet points are okay, too

#  - Typically a hyphen or asterisk is used for the bullet, preceded
#    by a single space, with blank lines in between, but conventions
#    vary here

# If you use an issue tracker, put references to them at the bottom,
# like this:

# Resolves: #123
# See also: #456, #789

gbolmier committed:
Fix average_precision_score tests errors

# More detailed explanatory text, if necessary. Wrap it to about 72
# characters or so. In some contexts, the first line is treated as the
# subject of the commit and the rest of the text as the body. The
# blank line separating the summary from the body is critical (unless
# you omit the body entirely); various tools like `log`, `shortlog`
# and `rebase` can get confused if you run the two together.

# Explain the problem that this commit is solving. Focus on why you
# are making this change as opposed to how (the code explains that).
# Are there side effects or other unintuitive consequences of this
# change? Here's the place to explain them.

# Further paragraphs come after blank lines.

#  - Bullet points are okay, too

#  - Typically a hyphen or asterisk is used for the bullet, preceded
#    by a single space, with blank lines in between, but conventions
#    vary here

# If you use an issue tracker, put references to them at the bottom,
# like this:

# Resolves: #123
# See also: #456, #789

gbolmier committed:
Make multiclass case handling more explicit in average_precision_score

gbolmier committed:
Add multiclass tests for average_precision_score

gbolmier committed:
Update user guide regarding average_precision_score multiclass support

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

gbolmier committed:
Add average_precision_score multiclass support entry in v0.24.rst change log

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

gbolmier committed:
Fix whats_new/0.24.rst

gbolmier committed:
Fix average_precision_score docstring

gbolmier committed:
Fix test_average_precision_score_pos_label_errors

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

gbolmier committed:
Merge remote-tracking branch 'upstream/master' into mean_average_precision_score

Commit:
Merge remote-tracking branch 'upstream/main' into mean_average_precision_score

Commit:
Merge remote-tracking branch 'upstream/main' into mean_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Update doc/modules/model_evaluation.rst

awinml committed:
Update Changelog for v1.2

awinml committed:
Update sklearn/metrics/_ranking.py

awinml committed:
Update sklearn/metrics/tests/test_common.py

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

glemaitre commented:
I am converting this PR to a draft until it is in a stage to be reviewed.
Giving a quick look, it seems that we will need to add a couple of tests as well.

awinml committed:
Add tests

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml commented:
@glemaitre I have added a test for the multiclass case. 
I am not able to figure out how to fix these failing tests. Can you please point me in the right direction?

```pytest sklearn/metrics/tests/test_classification.py```
The test case fails:
```python
___________ test_average_precision_score_score_non_binary_class _____________

    def test_average_precision_score_score_non_binary_class():
        y_true = [0, 2, 1, 1, 2, 0]
        y_score = np.array([[0.7, 0.2, 0.1],
                            [0.4, 0.3, 0.3],
                            [0.1, 0.8, 0.1],
                            [0.2, 0.3, 0.5],
                            [0.4, 0.4, 0.2],
                            [0.1, 0.2, 0.7],
                            ])
        err_msg = "multiclass format is not supported"
        with pytest.raises(ValueError, match=err_msg):
>           average_precision_score(y_true, y_score)
E           Failed: DID NOT RAISE <class 'ValueError'>

sklearn/metrics/tests/test_classification.py:364: Failed

```

The following test fails as well:
```python
pytest sklearn/metrics/_plot/tests/test_precision_recall_display.py
```

```python
____________________________ test_precision_recall_display_validation  _______________________________

pyplot = <module 'matplotlib.pyplot' from 'env6-sklearn/lib/python3.8/site-packages/matplotlib/pyplot.py'>

    def test_precision_recall_display_validation(pyplot):
        """Check that we raise the proper error when validating parameters."""
        X, y = make_classification(
            n_samples=100, n_informative=5, n_classes=5, random_state=0
        )
    
        with pytest.raises(NotFittedError):
            PrecisionRecallDisplay.from_estimator(SVC(), X, y)
    
        regressor = SVR().fit(X, y)
        y_pred_regressor = regressor.predict(X)
        classifier = SVC(probability=True).fit(X, y)
        y_pred_classifier = classifier.predict_proba(X)[:, -1]
    
        err_msg = "PrecisionRecallDisplay.from_estimator only supports classifiers"
        with pytest.raises(ValueError, match=err_msg):
            PrecisionRecallDisplay.from_estimator(regressor, X, y)
    
        err_msg = "Expected 'estimator' to be a binary classifier, but got SVC"
        with pytest.raises(ValueError, match=err_msg):
            PrecisionRecallDisplay.from_estimator(classifier, X, y)
    
        err_msg = "{} format is not supported"
        with pytest.raises(ValueError, match=err_msg.format("continuous")):
            # Force `y_true` to be seen as a regression problem
            PrecisionRecallDisplay.from_predictions(y + 0.5, y_pred_classifier, pos_label=1)
        with pytest.raises(ValueError, match=err_msg.format("multiclass")):
>           PrecisionRecallDisplay.from_predictions(y, y_pred_classifier, pos_label=1)

sklearn/metrics/_plot/tests/test_precision_recall_display.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'sklearn.metrics._plot.precision_recall_curve.PrecisionRecallDisplay'>
y_true = array([1, 1, 4, 4, 0, 2, 2, 0, 4, 4, 3, 2, 3, 4, 0, 0, 0, 0, 0, 4, 3, 1,
       3, 2, 3, 1, 0, 4, 3, 2, 0, 3, 2, 1, 3,...1,
       1, 3, 2, 2, 2, 1, 0, 1, 2, 4, 0, 0, 4, 3, 2, 0, 3, 0, 4, 4, 3, 0,
       1, 3, 4, 2, 1, 4, 1, 2, 3, 4, 2, 4])
y_pred = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])
sample_weight = None, pos_label = 1, name = None, ax = None, kwargs = {}
precision = array([0.2       , 0.2020202 , 0.20408163, 0.20618557, 0.20833333,
       0.21052632, 0.21276596, 0.21505376, 0.206521...       , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       1.        ])
recall = array([1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 1.  , 0.95, 0.95, 0.95,
       0.95, 0.95, 0.95, 0.95, 0.95, 0.95, 0.... , 0.  , 0.  , 0.  , 0.  ,
       0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  ,
       0.  , 0.  ])
_ = array([0.01204022, 0.01887368, 0.01946549, 0.02061938, 0.02178643,
       0.02218429, 0.02306837, 0.02458144, 0.026615...27, 0.67806319, 0.67822619, 0.67913536, 0.69509875,
       0.70273824, 0.70819639, 0.70905044, 0.71127943, 0.72038944])

    @classmethod
    def from_predictions(
        cls,
        y_true,
        y_pred,
        *,
        sample_weight=None,
        pos_label=None,
        name=None,
        ax=None,
        **kwargs,
    ):
        """Plot precision-recall curve given binary class predictions.
    
        Parameters
        ----------
        y_true : array-like of shape (n_samples,)
            True binary labels.
    
        y_pred : array-like of shape (n_samples,)
            Estimated probabilities or output of decision function.
    
        sample_weight : array-like of shape (n_samples,), default=None
            Sample weights.
    
        pos_label : str or int, default=None
            The class considered as the positive class when computing the
            precision and recall metrics.
    
        name : str, default=None
            Name for labeling curve. If `None`, name will be set to
            `"Classifier"`.
    
        ax : matplotlib axes, default=None
            Axes object to plot on. If `None`, a new figure and axes is created.
    
        **kwargs : dict
            Keyword arguments to be passed to matplotlib's `plot`.
    
        Returns
        -------
        display : :class:`~sklearn.metrics.PrecisionRecallDisplay`
    
        See Also
        --------
        PrecisionRecallDisplay.from_estimator : Plot precision-recall curve
            using an estimator.
    
        Notes
        -----
        The average precision (cf. :func:`~sklearn.metrics.average_precision`)
        in scikit-learn is computed without any interpolation. To be consistent
        with this metric, the precision-recall curve is plotted without any
        interpolation as well (step-wise style).
    
        You can change this style by passing the keyword argument
        `drawstyle="default"`. However, the curve will not be strictly
        consistent with the reported average precision.
    
        Examples
        --------
        >>> import matplotlib.pyplot as plt
        >>> from sklearn.datasets import make_classification
        >>> from sklearn.metrics import PrecisionRecallDisplay
        >>> from sklearn.model_selection import train_test_split
        >>> from sklearn.linear_model import LogisticRegression
        >>> X, y = make_classification(random_state=0)
        >>> X_train, X_test, y_train, y_test = train_test_split(
        ...         X, y, random_state=0)
        >>> clf = LogisticRegression()
        >>> clf.fit(X_train, y_train)
        LogisticRegression()
        >>> y_pred = clf.predict_proba(X_test)[:, 1]
        >>> PrecisionRecallDisplay.from_predictions(
        ...    y_test, y_pred)
        <...>
        >>> plt.show()
        """
        check_matplotlib_support(f"{cls.__name__}.from_predictions")
    
        check_consistent_length(y_true, y_pred, sample_weight)
        pos_label = _check_pos_label_consistency(pos_label, y_true)
    
        precision, recall, _ = precision_recall_curve(
            y_true, y_pred, pos_label=pos_label, sample_weight=sample_weight
        )
>       average_precision = average_precision_score(
            y_true, y_pred, pos_label=pos_label, sample_weight=sample_weight
        )

sklearn/metrics/_plot/precision_recall_curve.py:379: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

y_true = array([[0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]...      [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1]])
y_score = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])

    def average_precision_score(
        y_true, y_score, *, average="macro", pos_label=1, sample_weight=None
    ):
        """Compute average precision (AP) from prediction scores.
    
        AP summarizes a precision-recall curve as the weighted mean of precisions
        achieved at each threshold, with the increase in recall from the previous
        threshold used as the weight:
    
        .. math::
            \\text{AP} = \\sum_n (R_n - R_{n-1}) P_n
    
        where :math:`P_n` and :math:`R_n` are the precision and recall at the nth
        threshold [1]_. This implementation is not interpolated and is different
        from computing the area under the precision-recall curve with the
        trapezoidal rule, which uses linear interpolation and can be too
        optimistic.
    
        Read more in the :ref:`User Guide <precision_recall_f_measure_metrics>`.
    
        Parameters
        ----------
        y_true : ndarray of shape (n_samples,) or (n_samples, n_classes)
            True binary labels or binary label indicators.
    
        y_score : ndarray of shape (n_samples,) or (n_samples, n_classes)
            Target scores, can either be probability estimates of the positive
            class, confidence values, or non-thresholded measure of decisions
            (as returned by :term:`decision_function` on some classifiers).
    
        average : {'micro', 'samples', 'weighted', 'macro'} or None, \
                default='macro'
            If ``None``, the scores for each class are returned. Otherwise,
            this determines the type of averaging performed on the data:
    
            ``'micro'``:
                Calculate metrics globally by considering each element of the label
                indicator matrix as a label.
            ``'macro'``:
                Calculate metrics for each label, and find their unweighted
                mean.  This does not take label imbalance into account.
            ``'weighted'``:
                Calculate metrics for each label, and find their average, weighted
                by support (the number of true instances for each label).
            ``'samples'``:
                Calculate metrics for each instance, and find their average.
    
            Will be ignored when ``y_true`` is binary.
    
        pos_label : int or str, default=1
            The label of the positive class. Only applied to binary ``y_true``.
            For multiclass or multilabel-indicator ``y_true``, ``pos_label`` is
            fixed to 1.
    
        sample_weight : array-like of shape (n_samples,), default=None
            Sample weights.
    
        Returns
        -------
        average_precision : float
            Average precision score.
    
        See Also
        --------
        roc_auc_score : Compute the area under the ROC curve.
        precision_recall_curve : Compute precision-recall pairs for different
            probability thresholds.
    
        Notes
        -----
        .. versionchanged:: 0.19
          Instead of linearly interpolating between operating points, precisions
          are weighted by the change in recall since the last operating point.
    
        References
        ----------
        .. [1] `Wikipedia entry for the Average precision
               <https://en.wikipedia.org/w/index.php?title=Information_retrieval&
               oldid=793358396#Average_precision>`_
    
        Examples
        --------
        >>> import numpy as np
        >>> from sklearn.metrics import average_precision_score
        >>> y_true = np.array([0, 0, 1, 1])
        >>> y_scores = np.array([0.1, 0.4, 0.35, 0.8])
        >>> average_precision_score(y_true, y_scores)
        0.83...
        >>> y_true = np.array([0, 0, 1, 1, 2, 2])
        >>> y_scores = np.array([
        ...     [0.7, 0.2, 0.1],
        ...     [0.4, 0.3, 0.3],
        ...     [0.1, 0.8, 0.1],
        ...     [0.2, 0.3, 0.5],
        ...     [0.4, 0.4, 0.2],
        ...     [0.1, 0.2, 0.7],
        ... ])
        >>> average_precision_score(y_true, y_scores)
        0.77...
        """
    
        def _binary_uninterpolated_average_precision(
            y_true, y_score, pos_label=1, sample_weight=None
        ):
            precision, recall, _ = precision_recall_curve(
                y_true, y_score, pos_label=pos_label, sample_weight=sample_weight
            )
            # Return the step function integral
            # The following works because the last entry of precision is
            # guaranteed to be 1, as returned by precision_recall_curve
            return -np.sum(np.diff(recall) * np.array(precision)[:-1])
    
        y_type = type_of_target(y_true, input_name="y_true")
        # For multiclass we can reuse the multi-label case
        if y_type == "multiclass":
            y_true = label_binarize(y_true, classes=np.unique(y_true))
    
        if y_type in ["multiclass", "multilabel-indicator"] and pos_label != 1:
            raise ValueError(
                "Parameter pos_label is fixed to 1 for "
                "multiclass or multilabel-indicator y_true. Do not "
                "set pos_label or set pos_label to 1."
            )
        elif y_type == "binary":
            # Convert to Python primitive type to avoid NumPy type / Python str
            # comparison. See https://github.com/numpy/numpy/issues/6784
            present_labels = np.unique(y_true).tolist()
            if len(present_labels) == 2 and pos_label not in present_labels:
                raise ValueError(
                    f"pos_label={pos_label} is not a valid label. It should be "
                    f"one of {present_labels}"
                )
        average_precision = partial(
            _binary_uninterpolated_average_precision, pos_label=pos_label
        )
>       return _average_binary_score(
            average_precision, y_true, y_score, average, sample_weight=sample_weight
        )

sklearn/metrics/_ranking.py:247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

binary_metric = functools.partial(<function average_precision_score.<locals>._binary_uninterpolated_average_precision at 0x7fe37772ae50>, pos_label=1)
y_true = array([[0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]...      [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1]])
y_score = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])
average = 'macro', sample_weight = None

    def _average_binary_score(binary_metric, y_true, y_score, average, sample_weight=None):
        """Average a binary metric for multilabel classification.
    
        Parameters
        ----------
        y_true : array, shape = [n_samples] or [n_samples, n_classes]
            True binary labels in binary label indicators.
    
        y_score : array, shape = [n_samples] or [n_samples, n_classes]
            Target scores, can either be probability estimates of the positive
            class, confidence values, or binary decisions.
    
        average : {None, 'micro', 'macro', 'samples', 'weighted'}, default='macro'
            If ``None``, the scores for each class are returned. Otherwise,
            this determines the type of averaging performed on the data:
    
            ``'micro'``:
                Calculate metrics globally by considering each element of the label
                indicator matrix as a label.
            ``'macro'``:
                Calculate metrics for each label, and find their unweighted
                mean.  This does not take label imbalance into account.
            ``'weighted'``:
                Calculate metrics for each label, and find their average, weighted
                by support (the number of true instances for each label).
            ``'samples'``:
                Calculate metrics for each instance, and find their average.
    
            Will be ignored when ``y_true`` is binary.
    
        sample_weight : array-like of shape (n_samples,), default=None
            Sample weights.
    
        binary_metric : callable, returns shape [n_classes]
            The binary metric function to use.
    
        Returns
        -------
        score : float or array of shape [n_classes]
            If not ``None``, average the score, else return the score for each
            classes.
    
        """
        average_options = (None, "micro", "macro", "weighted", "samples")
        if average not in average_options:
            raise ValueError("average has to be one of {0}".format(average_options))
    
        y_type = type_of_target(y_true)
        if y_type not in ("binary", "multilabel-indicator"):
            raise ValueError("{0} format is not supported".format(y_type))
    
        if y_type == "binary":
            return binary_metric(y_true, y_score, sample_weight=sample_weight)
    
        check_consistent_length(y_true, y_score, sample_weight)
        y_true = check_array(y_true)
>       y_score = check_array(y_score)

sklearn/metrics/_base.py:79: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

array = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])
accept_sparse = False

    def check_array(
        array,
        accept_sparse=False,
        *,
        accept_large_sparse=True,
        dtype="numeric",
        order=None,
        copy=False,
        force_all_finite=True,
        ensure_2d=True,
        allow_nd=False,
        ensure_min_samples=1,
        ensure_min_features=1,
        estimator=None,
        input_name="",
    ):
    
        """Input validation on an array, list, sparse matrix or similar.
    
        By default, the input is checked to be a non-empty 2D array containing
        only finite values. If the dtype of the array is object, attempt
        converting to float, raising on failure.
    
        Parameters
        ----------
        array : object
            Input object to check / convert.
    
        accept_sparse : str, bool or list/tuple of str, default=False
            String[s] representing allowed sparse matrix formats, such as 'csc',
            'csr', etc. If the input is sparse but not in the allowed format,
            it will be converted to the first listed format. True allows the input
            to be any format. False means that a sparse matrix input will
            raise an error.
    
        accept_large_sparse : bool, default=True
            If a CSR, CSC, COO or BSR sparse matrix is supplied and accepted by
            accept_sparse, accept_large_sparse=False will cause it to be accepted
            only if its indices are stored with a 32-bit dtype.
    
            .. versionadded:: 0.20
    
        dtype : 'numeric', type, list of type or None, default='numeric'
            Data type of result. If None, the dtype of the input is preserved.
            If "numeric", dtype is preserved unless array.dtype is object.
            If dtype is a list of types, conversion on the first type is only
            performed if the dtype of the input is not in the list.
    
        order : {'F', 'C'} or None, default=None
            Whether an array will be forced to be fortran or c-style.
            When order is None (default), then if copy=False, nothing is ensured
            about the memory layout of the output array; otherwise (copy=True)
            the memory layout of the returned array is kept as close as possible
            to the original array.
    
        copy : bool, default=False
            Whether a forced copy will be triggered. If copy=False, a copy might
            be triggered by a conversion.
    
        force_all_finite : bool or 'allow-nan', default=True
            Whether to raise an error on np.inf, np.nan, pd.NA in array. The
            possibilities are:
    
            - True: Force all values of array to be finite.
            - False: accepts np.inf, np.nan, pd.NA in array.
            - 'allow-nan': accepts only np.nan and pd.NA values in array. Values
              cannot be infinite.
    
            .. versionadded:: 0.20
               ``force_all_finite`` accepts the string ``'allow-nan'``.
    
            .. versionchanged:: 0.23
               Accepts `pd.NA` and converts it into `np.nan`
    
        ensure_2d : bool, default=True
            Whether to raise a value error if array is not 2D.
    
        allow_nd : bool, default=False
            Whether to allow array.ndim > 2.
    
        ensure_min_samples : int, default=1
            Make sure that the array has a minimum number of samples in its first
            axis (rows for a 2D array). Setting to 0 disables this check.
    
        ensure_min_features : int, default=1
            Make sure that the 2D array has some minimum number of features
            (columns). The default value of 1 rejects empty datasets.
            This check is only enforced when the input data has effectively 2
            dimensions or is originally 1D and ``ensure_2d`` is True. Setting to 0
            disables this check.
    
        estimator : str or estimator instance, default=None
            If passed, include the name of the estimator in warning messages.
    
        input_name : str, default=""
            The data name used to construct the error message. In particular
            if `input_name` is "X" and the data has NaN values and
            allow_nan is False, the error message will link to the imputer
            documentation.
    
            .. versionadded:: 1.1.0
    
        Returns
        -------
        array_converted : object
            The converted and validated array.
        """
        if isinstance(array, np.matrix):
            raise TypeError(
                "np.matrix is not supported. Please convert to a numpy array with "
                "np.asarray. For more information see: "
                "https://numpy.org/doc/stable/reference/generated/numpy.matrix.html"
            )
    
        xp, is_array_api = get_namespace(array)
    
        # store reference to original array to check if copy is needed when
        # function returns
        array_orig = array
    
        # store whether originally we wanted numeric dtype
        dtype_numeric = isinstance(dtype, str) and dtype == "numeric"
    
        dtype_orig = getattr(array, "dtype", None)
        if not hasattr(dtype_orig, "kind"):
            # not a data type (e.g. a column named dtype in a pandas DataFrame)
            dtype_orig = None
    
        # check if the object contains several dtypes (typically a pandas
        # DataFrame), and store them. If not, store None.
        dtypes_orig = None
        pandas_requires_conversion = False
        if hasattr(array, "dtypes") and hasattr(array.dtypes, "__array__"):
            # throw warning if columns are sparse. If all columns are sparse, then
            # array.sparse exists and sparsity will be preserved (later).
            with suppress(ImportError):
                from pandas.api.types import is_sparse
    
                if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
                    warnings.warn(
                        "pandas.DataFrame with sparse columns found."
                        "It will be converted to a dense numpy array."
                    )
    
            dtypes_orig = list(array.dtypes)
            pandas_requires_conversion = any(
                _pandas_dtype_needs_early_conversion(i) for i in dtypes_orig
            )
            if all(isinstance(dtype_iter, np.dtype) for dtype_iter in dtypes_orig):
                dtype_orig = np.result_type(*dtypes_orig)
    
        if dtype_numeric:
            if dtype_orig is not None and dtype_orig.kind == "O":
                # if input is object, convert to float.
                dtype = xp.float64
            else:
                dtype = None
    
        if isinstance(dtype, (list, tuple)):
            if dtype_orig is not None and dtype_orig in dtype:
                # no dtype conversion required
                dtype = None
            else:
                # dtype conversion required. Let's select the first element of the
                # list of accepted types.
                dtype = dtype[0]
    
        if pandas_requires_conversion:
            # pandas dataframe requires conversion earlier to handle extension dtypes with
            # nans
            # Use the original dtype for conversion if dtype is None
            new_dtype = dtype_orig if dtype is None else dtype
            array = array.astype(new_dtype)
            # Since we converted here, we do not need to convert again later
            dtype = None
    
        if force_all_finite not in (True, False, "allow-nan"):
            raise ValueError(
                'force_all_finite should be a bool or "allow-nan". Got {!r} instead'.format(
                    force_all_finite
                )
            )
    
        estimator_name = _check_estimator_name(estimator)
        context = " by %s" % estimator_name if estimator is not None else ""
    
        # When all dataframe columns are sparse, convert to a sparse array
        if hasattr(array, "sparse") and array.ndim > 1:
            with suppress(ImportError):
                from pandas.api.types import is_sparse
    
                if array.dtypes.apply(is_sparse).all():
                    # DataFrame.sparse only supports `to_coo`
                    array = array.sparse.to_coo()
                    if array.dtype == np.dtype("object"):
                        unique_dtypes = set([dt.subtype.name for dt in array_orig.dtypes])
                        if len(unique_dtypes) > 1:
                            raise ValueError(
                                "Pandas DataFrame with mixed sparse extension arrays "
                                "generated a sparse matrix with object dtype which "
                                "can not be converted to a scipy sparse matrix."
                                "Sparse extension arrays should all have the same "
                                "numeric type."
                            )
    
        if sp.issparse(array):
            _ensure_no_complex_data(array)
            array = _ensure_sparse_format(
                array,
                accept_sparse=accept_sparse,
                dtype=dtype,
                copy=copy,
                force_all_finite=force_all_finite,
                accept_large_sparse=accept_large_sparse,
                estimator_name=estimator_name,
                input_name=input_name,
            )
        else:
            # If np.array(..) gives ComplexWarning, then we convert the warning
            # to an error. This is needed because specifying a non complex
            # dtype to the function converts complex to real dtype,
            # thereby passing the test made in the lines following the scope
            # of warnings context manager.
            with warnings.catch_warnings():
                try:
                    warnings.simplefilter("error", ComplexWarning)
                    if dtype is not None and np.dtype(dtype).kind in "iu":
                        # Conversion float -> int should not contain NaN or
                        # inf (numpy#14412). We cannot use casting='safe' because
                        # then conversion float -> int would be disallowed.
                        array = _asarray_with_order(array, order=order, xp=xp)
                        if array.dtype.kind == "f":
                            _assert_all_finite(
                                array,
                                allow_nan=False,
                                msg_dtype=dtype,
                                estimator_name=estimator_name,
                                input_name=input_name,
                            )
                        array = xp.astype(array, dtype, copy=False)
                    else:
                        array = _asarray_with_order(array, order=order, dtype=dtype, xp=xp)
                except ComplexWarning as complex_warning:
                    raise ValueError(
                        "Complex data not supported\n{}\n".format(array)
                    ) from complex_warning
    
            # It is possible that the np.array(..) gave no warning. This happens
            # when no dtype conversion happened, for example dtype = None. The
            # result is that np.array(..) produces an array of complex dtype
            # and we need to catch and raise exception for such cases.
            _ensure_no_complex_data(array)
    
            if ensure_2d:
                # If input is scalar raise error
                if array.ndim == 0:
                    raise ValueError(
                        "Expected 2D array, got scalar array instead:\narray={}.\n"
                        "Reshape your data either using array.reshape(-1, 1) if "
                        "your data has a single feature or array.reshape(1, -1) "
                        "if it contains a single sample.".format(array)
                    )
                # If input is 1D raise error
                if array.ndim == 1:
>                   raise ValueError(
                        "Expected 2D array, got 1D array instead:\narray={}.\n"
                        "Reshape your data either using array.reshape(-1, 1) if "
                        "your data has a single feature or array.reshape(1, -1) "
                        "if it contains a single sample.".format(array)
                    )
E                   ValueError: Expected 2D array, got 1D array instead:
E                   array=[0.10386782 0.0524558  0.58315465 0.45327993 0.07791555 0.15119987
E                    0.0709106  0.06298027 0.67374408 0.70905044 0.08237694 0.11805193
E                    0.08163128 0.69509875 0.01946549 0.0329531  0.06237162 0.02218429
E                    0.14665628 0.72038944 0.14230314 0.08449789 0.30416534 0.23390218
E                    0.04628601 0.05341511 0.05305951 0.37926655 0.07835889 0.08689533
E                    0.12168432 0.0675931  0.11652037 0.24721922 0.17082249 0.11446786
E                    0.53277934 0.1086098  0.09520099 0.08756231 0.1003601  0.07708606
E                    0.13534776 0.14357321 0.71127943 0.12894114 0.09235778 0.06134
E                    0.09545106 0.13578599 0.01204022 0.18879977 0.67913536 0.02306837
E                    0.03079178 0.05724336 0.70273824 0.02661554 0.02458144 0.07187062
E                    0.03780824 0.02178643 0.08596559 0.34725672 0.11121094 0.099966
E                    0.04077032 0.27920307 0.037281   0.10313433 0.02061938 0.11898652
E                    0.05598472 0.10160899 0.0292432  0.15811378 0.06328219 0.03216144
E                    0.40436357 0.056376   0.15223353 0.09977628 0.23798532 0.10161842
E                    0.57429405 0.70819639 0.13756695 0.05069074 0.10380559 0.07126013
E                    0.6092326  0.03780228 0.0870959  0.67806319 0.11936065 0.11523202
E                    0.54506286 0.67487927 0.01887368 0.67822619].
E                   Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.

sklearn/utils/validation.py:893: ValueError

During handling of the above exception, another exception occurred:

pyplot = <module 'matplotlib.pyplot' from '/home/ashwin/environments/env6-sklearn/lib/python3.8/site-packages/matplotlib/pyplot.py'>

    def test_precision_recall_display_validation(pyplot):
        """Check that we raise the proper error when validating parameters."""
        X, y = make_classification(
            n_samples=100, n_informative=5, n_classes=5, random_state=0
        )
    
        with pytest.raises(NotFittedError):
            PrecisionRecallDisplay.from_estimator(SVC(), X, y)
    
        regressor = SVR().fit(X, y)
        y_pred_regressor = regressor.predict(X)
        classifier = SVC(probability=True).fit(X, y)
        y_pred_classifier = classifier.predict_proba(X)[:, -1]
    
        err_msg = "PrecisionRecallDisplay.from_estimator only supports classifiers"
        with pytest.raises(ValueError, match=err_msg):
            PrecisionRecallDisplay.from_estimator(regressor, X, y)
    
        err_msg = "Expected 'estimator' to be a binary classifier, but got SVC"
        with pytest.raises(ValueError, match=err_msg):
            PrecisionRecallDisplay.from_estimator(classifier, X, y)
    
        err_msg = "{} format is not supported"
        with pytest.raises(ValueError, match=err_msg.format("continuous")):
            # Force `y_true` to be seen as a regression problem
            PrecisionRecallDisplay.from_predictions(y + 0.5, y_pred_classifier, pos_label=1)
        with pytest.raises(ValueError, match=err_msg.format("multiclass")):
>           PrecisionRecallDisplay.from_predictions(y, y_pred_classifier, pos_label=1)
E           AssertionError: Regex pattern 'multiclass format is not supported' does not match 'Expected 2D array, got 1D array instead:\narray=[0.10386782 0.0524558  0.58315465 0.45327993 0.07791555 0.15119987\n 0.0709106  0.06298027 0.67374408 0.70905044 0.08237694 0.11805193\n 0.08163128 0.69509875 0.01946549 0.0329531  0.06237162 0.02218429\n 0.14665628 0.72038944 0.14230314 0.08449789 0.30416534 0.23390218\n 0.04628601 0.05341511 0.05305951 0.37926655 0.07835889 0.08689533\n 0.12168432 0.0675931  0.11652037 0.24721922 0.17082249 0.11446786\n 0.53277934 0.1086098  0.09520099 0.08756231 0.1003601  0.07708606\n 0.13534776 0.14357321 0.71127943 0.12894114 0.09235778 0.06134\n 0.09545106 0.13578599 0.01204022 0.18879977 0.67913536 0.02306837\n 0.03079178 0.05724336 0.70273824 0.02661554 0.02458144 0.07187062\n 0.03780824 0.02178643 0.08596559 0.34725672 0.11121094 0.099966\n 0.04077032 0.27920307 0.037281   0.10313433 0.02061938 0.11898652\n 0.05598472 0.10160899 0.0292432  0.15811378 0.06328219 0.03216144\n 0.40436357 0.056376   0.15223353 0.09977628 0.23798532 0.10161842\n 0.57429405 0.70819639 0.13756695 0.05069074 0.10380559 0.07126013\n 0.6092326  0.03780228 0.0870959  0.67806319 0.11936065 0.11523202\n 0.54506286 0.67487927 0.01887368 0.67822619].\nReshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.'.

sklearn/metrics/_plot/tests/test_precision_recall_display.py:51: AssertionError

```

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

glemaitre commented:
Since you want to add support for multiclass, `test_average_precision_score_score_non_binary_class` is not valid anymore because it will work with multiclass. We can keep the test but make sure that it fails in multitarget case for instance.

awinml commented:
I feel it is better if we remove `test_average_precision_score_score_non_binary_class`. 

Any ideas as to how to fix `test_precision_recall_display_validation` failing?

As far as I can tell:
The code fails while computing `average_precision_score` for the input below: 
```python
y_true = array([[0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]...      [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1]])
y_score = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])
```

`_average_binary_score` in the `average_precision_score` is the cause of the failure:
```python
    def average_precision_score(
        y_true, y_score, *, average="macro", pos_label=1, sample_weight=None
    ):
        """Compute average precision (AP) from prediction scores.
                     ... (has been trimmed for readability)
        """
    
        def _binary_uninterpolated_average_precision(
            y_true, y_score, pos_label=1, sample_weight=None
        ):
            precision, recall, _ = precision_recall_curve(
                y_true, y_score, pos_label=pos_label, sample_weight=sample_weight
            )
            # Return the step function integral
            # The following works because the last entry of precision is
            # guaranteed to be 1, as returned by precision_recall_curve
            return -np.sum(np.diff(recall) * np.array(precision)[:-1])
    
        y_type = type_of_target(y_true, input_name="y_true")
        # For multiclass we can reuse the multi-label case
        if y_type == "multiclass":
            y_true = label_binarize(y_true, classes=np.unique(y_true))
    
        if y_type in ["multiclass", "multilabel-indicator"] and pos_label != 1:
            raise ValueError(
                "Parameter pos_label is fixed to 1 for "
                "multiclass or multilabel-indicator y_true. Do not "
                "set pos_label or set pos_label to 1."
            )
        elif y_type == "binary":
            # Convert to Python primitive type to avoid NumPy type / Python str
            # comparison. See https://github.com/numpy/numpy/issues/6784
            present_labels = np.unique(y_true).tolist()
            if len(present_labels) == 2 and pos_label not in present_labels:
                raise ValueError(
                    f"pos_label={pos_label} is not a valid label. It should be "
                    f"one of {present_labels}"
                )
        average_precision = partial(
            _binary_uninterpolated_average_precision, pos_label=pos_label
        )
>       return _average_binary_score(
            average_precision, y_true, y_score, average, sample_weight=sample_weight
        )

sklearn/metrics/_ranking.py:247: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

binary_metric = functools.partial(<function average_precision_score.<locals>._binary_uninterpolated_average_precision at 0x7fe37772ae50>, pos_label=1)
y_true = array([[0, 1, 0, 0, 0],
       [0, 1, 0, 0, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 0, 0, 1],
       [1, 0, 0, 0, 0]...      [0, 0, 1, 0, 0],
       [0, 0, 0, 1, 0],
       [0, 0, 0, 0, 1],
       [0, 0, 1, 0, 0],
       [0, 0, 0, 0, 1]])
y_score = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])
average = 'macro', sample_weight = None

```

`_average_binary_score` receives a `y_score` array which is 1D:

```python

    def _average_binary_score(binary_metric, y_true, y_score, average, sample_weight=None):
        """Average a binary metric for multilabel classification.
    
        Parameters
        ----------
        y_true : array, shape = [n_samples] or [n_samples, n_classes]
            True binary labels in binary label indicators.
    
        y_score : array, shape = [n_samples] or [n_samples, n_classes]
            Target scores, can either be probability estimates of the positive
            class, confidence values, or binary decisions.
    
        average : {None, 'micro', 'macro', 'samples', 'weighted'}, default='macro'
            If ``None``, the scores for each class are returned. Otherwise,
            this determines the type of averaging performed on the data:
    
            ``'micro'``:
                Calculate metrics globally by considering each element of the label
                indicator matrix as a label.
            ``'macro'``:
                Calculate metrics for each label, and find their unweighted
                mean.  This does not take label imbalance into account.
            ``'weighted'``:
                Calculate metrics for each label, and find their average, weighted
                by support (the number of true instances for each label).
            ``'samples'``:
                Calculate metrics for each instance, and find their average.
    
            Will be ignored when ``y_true`` is binary.
    
        sample_weight : array-like of shape (n_samples,), default=None
            Sample weights.
    
        binary_metric : callable, returns shape [n_classes]
            The binary metric function to use.
    
        Returns
        -------
        score : float or array of shape [n_classes]
            If not ``None``, average the score, else return the score for each
            classes.
    
        """
        average_options = (None, "micro", "macro", "weighted", "samples")
        if average not in average_options:
            raise ValueError("average has to be one of {0}".format(average_options))
    
        y_type = type_of_target(y_true)
        if y_type not in ("binary", "multilabel-indicator"):
            raise ValueError("{0} format is not supported".format(y_type))
    
        if y_type == "binary":
            return binary_metric(y_true, y_score, sample_weight=sample_weight)
    
        check_consistent_length(y_true, y_score, sample_weight)
        y_true = check_array(y_true)
>       y_score = check_array(y_score)

sklearn/metrics/_base.py:79: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

array = array([0.10386782, 0.0524558 , 0.58315465, 0.45327993, 0.07791555,
       0.15119987, 0.0709106 , 0.06298027, 0.673744...6 , 0.03780228, 0.0870959 , 0.67806319, 0.11936065,
       0.11523202, 0.54506286, 0.67487927, 0.01887368, 0.67822619])

```

This is because `y_true` is a 2D matrix of one-hot encoded classes. But `y_score` is a 1D array.
How should I modify the implementation of `average_precision_score` so it accounts for this?

Do I need to modify `_average_binary_score`? If so, how should I proceed?

glemaitre commented:
> I feel it is better if we remove test_average_precision_score_score_non_binary_class

I don't think so. We should be raising an error in this case and we need to make sure that we raise the proper error message and not something random.

glemaitre commented on sklearn/metrics/_plot/tests/test_precision_recall_display.py:
```diff
@@ -48,7 +48,7 @@ def test_precision_recall_display_validation(pyplot):
         PrecisionRecallDisplay.from_predictions(y + 0.5, y_pred_classifier, pos_label=1)
     with pytest.raises(ValueError, match=err_msg.format("multiclass")):
-        PrecisionRecallDisplay.from_predictions(y, y_pred_regressor, pos_label=1)
+        PrecisionRecallDisplay.from_predictions(y, y_pred_classifier, pos_label=1)
```
Apparently, this change is raising an error. We are expected to pass a 2D-array and we only pass a 1D array. So we should adapt the error message.

awinml replied:
Yes, I will change the error message.

awinml commented:
> > I feel it is better if we remove test_average_precision_score_score_non_binary_class
> 
> I don't think so. We should be raising an error in this case and we need to make sure that we raise the proper error message and not something random.

Okay. I will make sure it raises the proper error message.

awinml committed:
Update Error messages for tests of average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Update test_precision_recall_display.py

awinml committed:
Add tests for average_precision_score multiclass case

awinml commented:
@glemaitre I have added the required tests for the other cases while computing `average_precision_score`. Is there anything else that is needed to be added?

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -350,17 +350,49 @@ def test_precision_recall_f_ignored_labels():
-    y_true = rng.randint(0, 3, size=10)
-    err_msg = "multiclass format is not supported"
+def test_average_precision_score_non_binary_class():
+    # Test MultiClass MultiOuptut for average_precision_score
```
```suggestion
    """Test multiclass-multiouptut for `avergae_precision_score`."""
```

glemaitre commented on sklearn/metrics/tests/test_ranking.py:
```diff
@@ -1126,8 +1126,9 @@ def test_average_precision_score_pos_label_errors():
     y_true = np.array([[1, 0], [0, 1], [0, 1], [1, 0]])
```
could we make a separate case where we have mutliclass `y_true`, e.g. `y_true = [0, 1, 2, 0, 1, 2]`

awinml committed:
Update sklearn/metrics/tests/test_classification.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

awinml commented:
Yes, that can be done. I will create a separate test for the multiclass case.

awinml committed:
Update test_classification.py

awinml committed:
Add test for multiclass pos_label error

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Update docs

awinml commented:
@glemaitre I added the required test for checking incorrect pos_labels in the multiclass case. 
I also cleaned up some typos and added the change in the documentation and user guide for `average_precision_score".

awinml committed:
Update _ranking.py

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Fix Linting error

awinml commented:
@glemaitre I have fixed the linting errors. It passes the CI tests now.

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

glemaitre commented on doc/whats_new/v1.2.rst:
```diff
@@ -514,6 +514,11 @@ Changelog
   (`average="micro"`) for the One-vs-Rest multiclass case (`multi_class="ovr"`).
   :pr:`24338` by :user:`Arturo Amor <ArturoAmorQ>`.
+  
+- |Feature| :func:`metrics.average_precision_score` now supports the
+  multiclass case.
+  :pr:`17388` by :user:`Geoffrey Bolmier <gbolmier>` and
+  :pr:`24769` by :user:`Ashwin Mathur <awinml>`.
```
We will need to move this to 1.3 because the feature freeze for 1.2 has been done.

glemaitre commented on sklearn/metrics/_ranking.py:
```diff
@@ -213,17 +221,27 @@ def _binary_uninterpolated_average_precision(
-        if len(present_labels) == 2 and pos_label not in present_labels:
+
+    elif y_type == "multiclass":
+        y_true = label_binarize(y_true, classes=np.unique(y_true))
```
```suggestion
        y_true = label_binarize(y_true, classes=present_labels)
```

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -350,38 +350,99 @@ def test_precision_recall_f_ignored_labels():
+        average_precision_score(y_true, y_score, pos_label=2)
+
+
+def test_average_precision_score_multiclass_duplicate_values():
```
It looks like the same test as the one below. We just need to parameterize it and provide a different `y_true`/`y_pred`.

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -350,38 +350,99 @@ def test_precision_recall_f_ignored_labels():
     assert average_precision_score(y_true, y_score) == 1
 
 
+def test_average_precision_score_multiclass_tied_values():
```
Same as above.

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Update Changelog

awinml committed:
Update sklearn/metrics/_ranking.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

awinml committed:
Parameterize Tests

awinml commented:
@glemaitre The CI fails due to an unrelated issue.

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

glemaitre reviewed (approved):
Thanks @awinml LGTM

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml commented:
@glemaitre The PR has not been merged yet, is there something else that is required to be completed for it?

glemaitre commented:
We just need another reviewer to look at it.

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml commented:
Requesting Review: @jeremiedbb Would please kindly review this PR?

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml commented:
Requesting Review: @thomasjpfan  Would you please kindly review this PR?

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Fix linting in test_precision_recall_display.py

awinml commented:
@glemaitre This PR seems to have fallen off the radar. Any suggestions on who I should tag for the second review?

glemaitre commented:
It is tagged as "Waiting for the second reviewer" so I think it will be fine. I will add the 1.3 milestone as well. I don't want to aggressively request review because I assume that most of the reviewers have currently their own priorities.

Having the two labels will allow us to come to it before the release process.

awinml commented:
@glemaitre Thanks, appreciate the help.

thomasjpfan commented on sklearn/metrics/_ranking.py:
```diff
@@ -221,17 +229,27 @@ def _binary_uninterpolated_average_precision(
+
+    elif y_type == "multiclass":
+        y_true = label_binarize(y_true, classes=present_labels)
+        if pos_label not in present_labels:
```
I do not think `pos_label` makes sense for `multiclass`. For example, using `pos_label=2` would give zero:

```python
import numpy as np
from sklearn.metrics import average_precision_score
y_true = np.array([0, 0, 1, 1, 2, 2])
y_scores = np.array([
    [0.7, 0.2, 0.1],
    [0.4, 0.3, 0.3],
    [0.1, 0.8, 0.1],
    [0.2, 0.3, 0.5],
    [0.4, 0.4, 0.2],
    [0.1, 0.2, 0.7],
])

# This gives warnings about no positive class
average_precision_score(y_true, y_scores, pos_label=)
0.0
```

I prefer to error if `pos_label != 1`, like the multilabel-indicator case.

@glemaitre WDYT?

glemaitre replied:
Indeed, `pos_label` is only meaningful for the binary classification case.
I think this is reasonable to raise an error.

awinml replied:
Yes, `pos_label` only makes sense for the binary case. I will add the error for the `pos_label != 1` case.

thomasjpfan reviewed (commented):
Thank you for the PR @awinml !

awinml committed:
Raise error for pos_label != 1 multiclass case

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml committed:
Merge branch 'main' into multiclass_average_precision_score

awinml commented:
@glemaitre  The CI fails on linting. Running black on the files locally, does not produce any changes for me. Is there a specific flag that needs to be enabled now? I am using black-22.3.0.

awinml committed:
Fix linting

awinml commented:
@glemaitre Thanks! Adding the pre-commit hook fixed the problem.

thomasjpfan commented on sklearn/metrics/_ranking.py:
```diff
@@ -221,21 +229,33 @@ def _binary_uninterpolated_average_precision(
+                "Parameter pos_label is fixed to 1 for multiclass y_true. "
+                "Do not set pos_label or set pos_label to 1."
+            )
+        else:
+            y_true = label_binarize(y_true, classes=present_labels)
```
Nit: I prefer not to indent when there is an error check that exits early

```suggestion
        y_true = label_binarize(y_true, classes=present_labels)
```

thomasjpfan reviewed (approved):
I left a minor comment, otherwise this LGTM

awinml committed:
Update sklearn/metrics/_ranking.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
